iommu.c

```diff
@@ -35,18 +35,18 @@ u32 iommu_locate(void)
                       PCI_DEVFN(IOMMU_PCI_DEVICE, IOMMU_PCI_FUNCTION));
 }
 
-static void send_command(u64 *mmio_base, iommu_command_t cmd)
+static void send_command(u64 *mmio_base, iommu_command_t cmd, u32 *idx)
 {
-    u32 cmd_ptr = mmio_base[IOMMU_MMIO_COMMAND_BUF_TAIL] >> 4;
-    command_buf[cmd_ptr++] = cmd;
+    u32 cmd_ptr = mmio_base[IOMMU_MMIO_COMMAND_BUF_TAIL];
+    command_buf[(*idx)++] = cmd;
     smp_wmb();
-    mmio_base[IOMMU_MMIO_COMMAND_BUF_TAIL] = (cmd_ptr << 4);
+    mmio_base[IOMMU_MMIO_COMMAND_BUF_TAIL] = cmd_ptr + sizeof(iommu_command_t);
```

Comment:
That looks reasonable. What do you mean by "check where the increment wants to happen"?

Reply:
Actually looking at it, I don't think that will work. The index has to be reset on every call to iommu_load_device_table() because there are only two slots in the command_buf. That is why I was passing a pointer to the index.

Reply:
Shouldn't value written to IOMMU_MMIO_COMMAND_BUF_TAIL be masked to clamp it to size of buffer?

Reply:
Actually yea there is something more missing there. The tail pointer is not an address but an offset.

Reply:
that will go wrong depending on the alignment of cmd buff.

It wants to be `= *idx * sizeof(cmd)` I think

Reply:
Almost, but that only works when command_buf doesn't cross a page boundary.  The tail pointer is a byte offset from the beginning of the 4k page that command_buf starts in.  So the actual calculation is `_u(&command_buf[*idx]) - (_u(&command_buf) & ~0xfff)` but maybe it's just easier to store the tail pointer calculated to begin with, and add one to it.

Reply:
Urrg yea you are right. I keep forgetting the buffer is 8k. I think we run into the index reset issue with the storing of the initial tail value too. Gotta think about it.